arangod/RestServer/DatabaseFeature.cpp

```diff
@@ -291,6 +294,165 @@ void DatabaseManagerThread::run() {
+  std::string testFilePath = FileUtils::buildFilename(
+      databasePathFeature.directory().c_str(), "TestBalloonFileIOHeartbeat");
+  std::string testFileContent = "This is just an I/O test.\n";
+
+  LOG_TOPIC("66665", INFO, Logger::ENGINES)
+      << "IOHeartbeatThread: running with checkInterval= " << checkInterval()
+      << " ...";
+
+  unsigned int counter = checkInterval();
+
+  while (true) {
+    try {  // protect thread against any exceptions
+      if (isStopping()) {
+        // done
+        break;
+      }
+
+      if (--counter == 0) {
+        LOG_TOPIC("66659", DEBUG, Logger::ENGINES)
+            << "IOHeartbeat: testing to write/read/remove " << testFilePath;
+        counter = checkInterval();
+        // We simply write a file and sync it to disk in the database
+        // directory and then read it and then delete it again:
+        auto start1 = std::chrono::steady_clock::now();
+        bool trouble = false;
+        try {
+          FileUtils::spit(testFilePath, testFileContent, true);
+        } catch (std::exception const& exc) {
+          ++_writeFailures;
+          LOG_TOPIC("66663", INFO, Logger::ENGINES)
+              << "IOHeartbeat: exception when writing test file: "
+              << exc.what();
+          trouble = true;
+        }
+        auto finish = std::chrono::steady_clock::now();
+        std::chrono::duration<double> dur = finish - start1;
+        bool delayed = dur > std::chrono::seconds(1);
+        if (trouble || delayed) {
+          if (delayed) {
+            ++_writeDelays;
+          }
+          LOG_TOPIC("66662", INFO, Logger::ENGINES)
+              << "IOHeartbeat: trying to write test file took "
+              << std::chrono::duration_cast<std::chrono::microseconds>(dur)
+                     .count()
+              << " milliseconds.";
+        }
+
+        // Read the file if we can reasonably assume it is there:
+        if (!trouble) {
+          auto start = std::chrono::steady_clock::now();
+          try {
+            std::string content = FileUtils::slurp(testFilePath);
+          } catch (std::exception const& exc) {
+            LOG_TOPIC("66661", INFO, Logger::ENGINES)
+                << "IOHeartbeat: exception when reading test file: "
+                << exc.what();
+            trouble = true;
+          }
+          auto finish = std::chrono::steady_clock::now();
+          std::chrono::duration<double> dur = finish - start;
+          bool delayed = dur > std::chrono::seconds(1);
+          if (trouble || delayed) {
+            if (delayed) {
+              ++_readDelays;
+            }
+            LOG_TOPIC("66669", INFO, Logger::ENGINES)
+                << "IOHeartbeat: trying to read test file took "
+                << std::chrono::duration_cast<std::chrono::microseconds>(dur)
+                       .count()
+                << " milliseconds.";
+          }
+        }
+
+        // And remove it again:
```

Comment:
This will be executed even if the initial writing of the file failed. The whole remove attempt should probably be moved into an if clause and only be executed if the initial write operation did not set `trouble` to `true`.

Reply:
One failure case is that the file could not be written because of permissions (since it is already there, but unwritable). Sometimes then the removal would work, that is why I had originally executed this in any case. But I have moved the removal into the if statement and only remove it if it has worked to create it.